SimG4Core/Generators/src/Generator3.cc

```diff
@@ -0,0 +1,585 @@
+        zimpact = (pz > 0.0) ? Z_hector : -Z_hector;
+        double del = (zimpact - z1) / pz;
+        ximpact += del * px;
+        yimpact += del * py;
+      }
+      double rimpact2 = ximpact * ximpact + yimpact * yimpact;
+
+      if (verbose > 2)
+        LogDebug("SimG4CoreGenerator3") << "Processing GenParticle barcode= " << pitr->id() << " pdg= " << pdg
+                                        << " status= " << pitr->status() << " st= " << status
+                                        << " rimpact(cm)= " << std::sqrt(rimpact2) / CLHEP::cm
+                                        << " zimpact(cm)= " << zimpact / CLHEP::cm << " ptot(GeV)= " << ptot
+                                        << " pz(GeV)= " << pz;
+
+      // Particles of status 1 trasnported along the beam pipe
+      // HECTOR transport of protons are done in corresponding PPS producer
+      if (1 == status && std::abs(zimpact) >= Z_hector && rimpact2 <= theDecRCut2) {
+        // very forward n, nbar, gamma are allowed
+        toBeAdded = (2112 == std::abs(pdg) || 22 == pdg);
+        if (verbose > 1) {
+          edm::LogVerbatim("SimG4CoreGenerator3")
+              << "GenParticle barcode = " << pitr->id() << " very forward; to be added: " << toBeAdded;
+        }
+      } else {
+        // Standard case: particles not decayed by the generator and not forward
+        if (1 == status && (std::abs(zimpact) < Z_hector || rimpact2 > theDecRCut2)) {
+          // Ptot cut for all particles
+          if (fPCuts && (ptot < theMinPCut || ptot > theMaxPCut)) {
+            continue;
+          }
+          // phi cut is applied mainly for particle gun
+          if (fPhiCuts) {
+            double phi = p.phi();
+            if (phi < theMinPhiCut || phi > theMaxPhiCut) {
+              continue;
+            }
+          }
+          // eta cut is applied if position of the decay
+          // is within vacuum chamber and limited in Z
+          if (fEtaCuts) {
+            // eta cut
+            double xi = x1;
+            double yi = y1;
+            double zi = z1;
+
+            // can be propagated along Z
+            if (std::abs(pz) >= minTan * ptot) {
+              if ((zi >= Z_lmax) & (pz < 0.0)) {
+                zi = Z_lmax;
+              } else if ((zi <= Z_lmin) & (pz > 0.0)) {
+                zi = Z_lmin;
+              } else {
+                if (pz > 0) {
+                  zi = Z_lmax;
+                } else {
+                  zi = Z_lmin;
+                }
+              }
+              double del = (zi - z1) / pz;
+              xi += del * px;
+              yi += del * py;
+            }
+            // check eta cut
+            if ((zi >= Z_lmin) & (zi <= Z_lmax) & (xi * xi + yi * yi < theDecRCut2)) {
+              continue;
+            }
+          }
+          const HepMC3::GenParticle *ppointer = pitr.get();
+          if (fLumiFilter && !fLumiFilter->isGoodForLumiMonitor(ppointer)) {  // MK: this function is always true
```

Comment:
Can you comment on the purpose of this? `isGoodForLumiMonitor` is indeed always true

Reply:
This comes from the class Generator, written by somebody else for HepMC, time ago.  May be there is a possibility to provide another, non-trivial function for special cases?

Reply:
Yeah but it's ok to do that in another PR if you prefer